Traffic api

Succo opened the pull request:
This is my take at https://github.com/google/go-github/issues/417

Critics welcome.

shurcooL commented on github/github.go:
```diff
@@ -84,6 +84,9 @@ const (
 	mediaTypePagesPreview = "application/vnd.github.mister-fantastic-preview+json"
+
+	// https://developer.github.com/v3/repos/traffic/
+	mediaTypeTrafficPreview = "application/vnd.github.spiderman-preview"
```
There should be a `+json` suffix, see similar lines above.

For more context, see https://github.com/google/go-github/pull/337#discussion_r66712727.

shurcooL commented on github/repos_traffic.go:
```diff
@@ -0,0 +1,167 @@
+// This is necessary becausegithub uses unix timestamp and not YYYY-MM-DDTHH:MM:SSZ
+type Time struct {
+	time.Time
+}
```
Can [`github.Timestamp`](https://godoc.org/github.com/google/go-github/github#Timestamp) be used, or is this different?

Succo replied:
Yeah I hadn't seen it.

willnorris replied:
(moving my comment from elsewhere, since I just saw this discussion...)

I do worry about this type causing confusion given the existing `Timestamp` type. Right now it handles seconds, but using some pretty basic heuristics, we should be able to determine if a given value is intended to be seconds or milliseconds. Something like:

```go
if i > 1e9 {
    i /= 1000
}
```

That'll at least work for dates up to the year 2286. What do you think?

willnorris replied:
Although I guess this approach also causes problems with times before 1970-01-12 that are expressed in milliseconds (since they'd be interpreted as seconds and be about 30 years off).  Not sure how likely it is to have dates in that range in the traffic API though.  

Another option, which is *much* cleaner, but requires a bit more work is to support some basic tag options.  So you could have:

```go
struct {
    Timestamp *Timestamp `json:"timestamp,omitempty" precision:"millisecond"`
}
```

or something.  Though I'm pretty sure you don't have access to the struct tags inside of `Timestamp.UnmarshalJSON`, so a lot would have to be reworked.  Almost certainly not worth the trouble.

Succo replied:
I'm pretty sure some kind of workaround would be doable using "smart" Unmarshalling and estimation on the range of the timestamp (comparing it to `time.Now().Unix()` maybe) but that seems the kind of risky workaround that could surprise us later.

I tend to prefer clarity and safety to brevity, but as you want.

willnorris replied:
Other alternative is just to rename the type for the time being (`TrafficTime` ? though that's still kinda terrible in other ways) so it doesn't get confused as being a general purpose time type.

willnorris replied:
Or I guess `TimestampMS` or some variant like that might be better since there is nothing actually specific to the Traffic API about this.

Succo replied:
That makes sense. I renamed it.

shurcooL commented on github/repos_traffic.go:
```diff
@@ -0,0 +1,167 @@
+	Uniques *int         `json:"uniques,omitempty"`
+}
+
+// Clones represent information about clones on the last 14 days
```
This sentence is missing a period at the end. Also applies to most comments above in this file.

shurcooL commented on github/github.go:
```diff
@@ -144,6 +147,11 @@ type UploadOptions struct {
 
+// BreakdownOptions specifies the parameters to methods that support breakdown per day or week.
+type BreakdownOptions struct {
+	Per string `url:"per,omitempty"`
```
I think it'll be very helpful to document this `Per` parameter with something like this:

```
// Can be one of: day, week. Default: day.
```

For example, see https://github.com/google/go-github/blob/b9698164669e386680621e2006cab02271257611/github/repos.go#L113-L115.

shurcooL commented:
Thank you for contributing @Succo!

I've left a few inline comments as my initial review, please take a look and respond.

This is a great start.

parkr commented on github/github.go:
```diff
@@ -144,6 +147,11 @@ type UploadOptions struct {
 }
 
+// BreakdownOptions specifies the parameters to methods that support breakdown per day or week.
+type BreakdownOptions struct {
```
Usually these options are scoped out to their method if only used once, e.g. TrafficBreakdownOptions. See OrgMembershipOptions, for example. But I defer to the real maintainers of this project for a second. :smile:

Succo commented:
Updated the PR. I think the PR ran on a previous version of the commit, before I amended it with gofmt.

Not sure how to rerun it. (But I added gofmt on save to my IDE, so should not happen again).

shurcooL commented:
@Succo, would you like to work on adding tests for the new code?

You can see a good example of tests added in PR [#393](https://github.com/google/go-github/pull/393/files).

Also see an example PR [#373](https://github.com/google/go-github/pull/373/files) which added support forrepository invitations API, and also added tests. You can use that as a reference.

But really, any of the _test.go files in this package serve as good examples of tests.

Succo commented:
@shurcooL I added test based on the example you gave me.

willnorris commented:
(while @shurcooL is working with you on the code review, we do still need a [CLA](https://cla.developers.google.com/) from you before we can merge this.)

Succo commented:
@willnorris, I know, I'm working on getting access back to my google account.

Succo commented:
ok, should be good now for the CLA.

shurcooL commented on github/activity_events_test.go:
```diff
@@ -229,7 +229,6 @@ func TestActivityService_ListEventsReceivedByUser_all(t *testing.T) {
 func TestActivityService_ListEventsReceivedByUser_publicOnly(t *testing.T) {
 	setup()
 	defer teardown()
-
```
The change on this line looks unintentional.

Succo replied:
My bad, amended the commit.

willnorris commented on github/github.go:
```diff
@@ -144,6 +147,12 @@ type UploadOptions struct {
 	Name string `url:"name,omitempty"`
 }
 
+// TrafficBreakdownOptions specifies the parameters to methods that support breakdown per day or week.
```
this should live in `repos_traffic.go` alongside the methods that use it.

willnorris commented on github/repos_traffic.go:
```diff
@@ -0,0 +1,167 @@
+// Copyright 2013 The go-github AUTHORS. All rights reserved.
```
use current year.  same in repos_traffic_test.go

willnorris commented on github/repos_traffic.go:
```diff
@@ -0,0 +1,167 @@
+// Time represents a time stamp as in a datapoint.
+//
+// This is necessary because github uses milliseconds timestamp.
+type Time struct {
```
rather than creating a new type, perhaps we could modify https://godoc.org/github.com/google/go-github/github#Timestamp to handle millisecond precision times.  Right now it handles seconds, but using some pretty basic heuristics, we should be able to determine if a given value is intended to be seconds or milliseconds.  Something like:

```go
if i > 1e9 {
    i /= 1000
}
```

That'll at least work for dates up to the year 2286.  What do you think?

willnorris commented on github/repos_traffic.go:
```diff
@@ -0,0 +1,167 @@
+	Uniques  *int    `json:"uniques,omitempty"`
+}
+
+// Path represent a referrer information.
```
copy/paste error from Referrer type above

willnorris commented on github/repos_traffic.go:
```diff
@@ -0,0 +1,167 @@
+}
+
+// Path represent a referrer information.
+type Path struct {
```
I worry a little about this name not being specific enough, given that file paths are used in a number of places throughout the library.  So maybe rename to `TrafficPath` or update the godoc to be very clear what this is referring to.

willnorris commented on github/repos_traffic.go:
```diff
@@ -0,0 +1,167 @@
+}
+
+// Datapoint represent a view in views on clones.
+type Datapoint struct {
```
maybe `TrafficView` here and `TrafficViews` below?  (I'm not really loving that myself, but these are awfully general names for such a specific part of the API)

willnorris replied:
did this comment get overlooked?

Succo replied:
Oh yeah sorry.

I renamed it, although it make it a bit less clearer for it's use in `TrafficClones`, not really inspired for a better name.

willnorris replied:
ah, fair point.  I didn't realize it got used there also.  `TrafficData`, or even leave the "datapoint" like you had... `TrafficDatapoint`?

Succo replied:
Wen't with TrafficData to avoid having a name too long.

willnorris commented on github/repos_traffic.go:
```diff
@@ -0,0 +1,167 @@
+// ListReferrers list the top 10 referrers over the last 14 days.
+//
+// GitHub API docs: https://developer.github.com/v3/repos/traffic/#list-referrers
+func (s *RepositoriesService) ListReferrers(owner, repo string) ([]*Referrer, *Response, error) {
```
Same concern as above, but now on the methods.  I wonder if this should be `ListTrafficReferrers` (and similar naming style for methods below).  The fact that the URL for this is `/repos/o/r/traffic/popular/referrers` rather than simply `/repos/o/r/referrers` is a good indicator that maybe we should rename this.

willnorris commented on github/repos_traffic_test.go:
```diff
@@ -0,0 +1,148 @@
+	setup()
+	defer teardown()
+
+	mux.HandleFunc("/repos/1/2/traffic/popular/referrers", func(w http.ResponseWriter, r *http.Request) {
```
nit: as a matter of convention, we pretty much always use "o" and "r" for placeholder values to represent "owner" and "repo".  So, "/repos/o/r/traffic/popular/referrers"

Succo commented:
@willnorris I went with a new naming scheme using traffic pretty much everywhere and clarifying the comments.

willnorris commented:
This looks great.  Thanks for pushing through all the rounds of review! :) :+1:
